Improve password checking

karthikprabhuvinod opened the pull request:
Signed-off-by: Karthik Prabhu Vinod <karthik.prabhu.vinod@intel.com>

Fixes Issue: #398 

Changes proposed in this pull request:
- Leveraging the use of cracklib-check to make sure passwords comply with some standard during creation
- We still maintain the min, max character len policy as before as cracklib-check allows for passwords to be smaller than 8 characters
- also implements from the issue the following:

- [X] disallow common words (or full dictionary if available)
- [X] disallow repeated characters
- [X] Hard disk encryption passphrase also checks for cracklib
- [X] cl-installler --genpass also checks for cracklib

karthikprabhuvinod commented:
So the password has to pass the following baptims of fire (Order intended):

min of 8 char
max of 255 charcacters
Unfamiliar to username OR login (The password should not more 35 percent similiar to either of those)
Atleast one special character
Atleast one number
Pass through shell based cracklib-check

karthikprabhuvinod commented:
Looks like the Make check functionality is broken on travis as it passes on travis and brakes locally

karthikprabhuvinod commented:
added a large bunch of test

karthikprabhuvinod commented:
@mdhorn  I was looking at the cracklib github, https://github.com/cracklib/cracklib/blob/8e921bfbc9ea3f1bbd6695eefa5f81f1e878d069/src/lib/crack.h#L15 ...looks like they look at currently logged in user. 

At this point, I will drop the user password attribute similarity check . Also I think the PR is ready to be reviewed.

mdhorn commented on locale/en_US/LC_MESSAGES/clr-installer.po:
```diff
@@ -239,8 +239,8 @@ msgstr "RESCAN MEDIA"
 
-msgid "Username must contain only numbers, letters, commas, - or _"
-msgstr "Username must contain only numbers, letters, commas, - or _"
+msgid "Username must contain only numbers, letters, commas, - or _ "
```
There is whitespace at the end of this string.

karthikprabhuvinod replied:
Actually My editor kept complaining that a string cant end with a . or _. So i added this

mdhorn replied:
But the editor is wrong.  😃

karthikprabhuvinod replied:
Done

mdhorn commented on locale/en_US/LC_MESSAGES/clr-installer.po:
```diff
@@ -253,8 +253,8 @@ msgstr "Login is required"
 
-msgid "Login must contain only numbers, letters, -, _ or ."
-msgstr "Login must contain only numbers, letters, -, _ or ."
+msgid "Login must contain only numbers, letters, -, _ or . "
```
There is whitespace at the end of this string.

karthikprabhuvinod replied:
Same here

karthikprabhuvinod replied:
So the linting error was that I cant a sentence starting with a uppercase char and ending with a full stop. so I interchanged positions of . with _

mdhorn commented on locale/en_US/LC_MESSAGES/clr-installer.po:
```diff
@@ -809,3 +809,23 @@ msgstr "Network Required"
+msgid "Password is too similar to the user attribute: %s"
+msgstr "Password is too similar to the user attribute: %s"
+
+msgid "Password should contain atleast one (special character, digit & Uppercase)"
```
atleast should be at least
```suggestion
msgid "Password should contain at least one (special character, digit & Uppercase)"
```

karthikprabhuvinod replied:
Done

mdhorn commented on locale/en_US/LC_MESSAGES/clr-installer.po:
```diff
@@ -809,3 +809,23 @@ msgstr "Network Required"
+msgstr "Password is based on a dictionary word"
+
+msgid "it is based on a (reversed) dictionary word"
+msgstr "Password is based on a (reversed) dictionary word"
```
Need a carriage return at the end of this line.

karthikprabhuvinod replied:
Okay Thanks...Done and ensured in all PO files

mdhorn commented on user/user_validator.go:
```diff
@@ -0,0 +1,147 @@
+)
+
+// NewValidator creates/allocates a new user validation
+func NewValidator(login string,
```
Since we only return nil for the error, and all of the calls to NewValidator always ignore the error, please just drop returning an error and only return the validator.

karthikprabhuvinod replied:
Makes sense..Since we dont perform any checks at object creation..this can b removed

karthikprabhuvinod replied:
Done

mdhorn commented on user/user_validator_test.go:
```diff
@@ -0,0 +1,169 @@
+}
+
+const (
+	cracklibpath = "/usr/bin/cracklib-check"
```
Please move this const into user_validator and reference the const in place of the same string in that code.

karthikprabhuvinod replied:
Done

karthikprabhuvinod commented on Makefile:
```diff
@@ -35,8 +35,11 @@ LOCAL_GOPATH := ${CURDIR}/.gopath
 
-CLR_INSTALLER_TEST_HTTP_PORT ?= 8181
 
+# Required for tests to work well 
```
Done

karthikprabhuvinod commented on Makefile:
```diff
@@ -35,8 +35,11 @@ LOCAL_GOPATH := ${CURDIR}/.gopath
 
+# Required for tests to work well 
+export CLR_INSTALLER_LOCALE_DIR := $(top_srcdir)/locale
+export CLR_INSTALLER_ISO_TEMPLATE_DIR := $(top_srcdir)/iso_templates
```
Done

karthikprabhuvinod commented on go.mod:
```diff
@@ -7,13 +7,15 @@ require (
 	github.com/VladimirMarkelov/clui v1.2.1
 	github.com/atotto/clipboard v0.1.2 // indirect
 	github.com/coreos/go-systemd/v22 v22.0.0
+	github.com/deckarep/golang-set v1.7.1 // indirect
```
Done

karthikprabhuvinod replied:
Sorry I realized for some reason it didnt clean, I ran again  run go mod verify and go mod vendor and it cleaned up

karthikprabhuvinod commented on go.mod:
```diff
@@ -7,13 +7,15 @@ require (
 	github.com/kr/pretty v0.2.0 // indirect
 	github.com/kr/text v0.2.0 // indirect
 	github.com/leonelquinteros/gotext v1.4.0
+	github.com/masatana/go-textdistance v0.0.0-20191005053614-738b0edac985 // indirect
```
Done

karthikprabhuvinod commented on go.sum:
```diff
@@ -8,6 +8,8 @@ github.com/coreos/go-systemd/v22 v22.0.0 h1:XJIw/+VlJ+87J+doOxznsAWIdmWuViOVhkQa
 github.com/coreos/go-systemd/v22 v22.0.0/go.mod h1:xO0FLkIi5MaZafQlIrOotqXZ90ih+1atmu1JpKERPPk=
 github.com/creack/pty v1.1.9 h1:uDmaGzcdjhF4i/plgjmEsriH11Y0o7RKapEf/LDaM3w=
 github.com/creack/pty v1.1.9/go.mod h1:oKZEueFk5CKHvIhNR5MUki03XCEU+Q6VDXinZuGJ33E=
+github.com/deckarep/golang-set v1.7.1 h1:SCQV0S6gTtp6itiFrTqI+pfmJ4LN85S1YzhDf9rTHJQ=
```
Done

karthikprabhuvinod commented on go.sum:
```diff
@@ -8,6 +8,8 @@ github.com/coreos/go-systemd/v22 v22.0.0 h1:XJIw/+VlJ+87J+doOxznsAWIdmWuViOVhkQa
 github.com/creack/pty v1.1.9 h1:uDmaGzcdjhF4i/plgjmEsriH11Y0o7RKapEf/LDaM3w=
 github.com/creack/pty v1.1.9/go.mod h1:oKZEueFk5CKHvIhNR5MUki03XCEU+Q6VDXinZuGJ33E=
+github.com/deckarep/golang-set v1.7.1 h1:SCQV0S6gTtp6itiFrTqI+pfmJ4LN85S1YzhDf9rTHJQ=
+github.com/deckarep/golang-set v1.7.1/go.mod h1:93vsz/8Wt4joVM7c2AVqh+YRMiUSc14yDtF28KmMOgQ=
```
Done

karthikprabhuvinod commented on go.sum:
```diff
@@ -27,6 +29,8 @@ github.com/kr/text v0.2.0 h1:5Nx0Ya0ZqY2ygV366QzturHI13Jq95ApcVaJBhpS+AY=
 github.com/kr/text v0.2.0/go.mod h1:eLer722TekiGuMkidMxC/pM04lWEeraHUUmBw8l2grE=
 github.com/leonelquinteros/gotext v1.4.0 h1:2NHPCto5IoMXbrT0bldPrxj0qM5asOCwtb1aUQZ1tys=
 github.com/leonelquinteros/gotext v1.4.0/go.mod h1:yZGXREmoGTtBvZHNcc+Yfug49G/2spuF/i/Qlsvz1Us=
+github.com/masatana/go-textdistance v0.0.0-20191005053614-738b0edac985 h1:Pz8zZjVRvKxISYimNzLGnzSNl5hYXFSN80FPQ+qt1HE=
```
Done

karthikprabhuvinod commented on go.sum:
```diff
@@ -27,6 +29,8 @@ github.com/kr/text v0.2.0 h1:5Nx0Ya0ZqY2ygV366QzturHI13Jq95ApcVaJBhpS+AY=
 github.com/leonelquinteros/gotext v1.4.0 h1:2NHPCto5IoMXbrT0bldPrxj0qM5asOCwtb1aUQZ1tys=
 github.com/leonelquinteros/gotext v1.4.0/go.mod h1:yZGXREmoGTtBvZHNcc+Yfug49G/2spuF/i/Qlsvz1Us=
+github.com/masatana/go-textdistance v0.0.0-20191005053614-738b0edac985 h1:Pz8zZjVRvKxISYimNzLGnzSNl5hYXFSN80FPQ+qt1HE=
+github.com/masatana/go-textdistance v0.0.0-20191005053614-738b0edac985/go.mod h1:1nU7rI+iBPtzc9ZKOqeQacD290rA0wcJLu5AtOSBBPw=
```
Done

mdhorn commented on clr-installer/main.go:
```diff
@@ -138,6 +139,11 @@ func execute(options args.Args) error {
 
 	if options.PamSalt != "" {
+		if status, err := user.IsValidPassword(options.PamSalt); !status {
+			fmt.Printf("%s\n", err)
```
We should not exit, but instead, return an error
```suggestion
			return err
```

karthikprabhuvinod replied:
okay Thanks. I am going to add a extra carriage return to separate the error from the output of the password. Can be a little hard to see

mdhorn commented on user/user_validator.go:
```diff
@@ -0,0 +1,148 @@
+var (
+	usernameExp  = regexp.MustCompile("^([a-zA-Z]+[0-9a-zA-Z-_ ,'.]*|)$")
+	loginExp     = regexp.MustCompile("^[a-zA-Z]+[0-9a-zA-Z-_.]*$")
+	cracklibpath = "/usr/bin/cracklib-check"
```
This should be const as it was in the other file.
Also, for consistency, please rename it to cracklibPath

karthikprabhuvinod replied:
Done

karthikprabhuvinod commented:
@mdhorn Addressed all concerns! Please take a look when you can

mdhorn commented on locale/en_US/LC_MESSAGES/clr-installer.po:
```diff
@@ -812,3 +812,23 @@ msgstr "Network Required"
 msgstr "Installation Steps Complete"
+
+#, c-format
+msgid "Password is too similar to the user attribute: %s"
```
Is this translation still used?  I can't find a reference to it.

karthikprabhuvinod replied:
Thanks for finding this..removed it

karthikprabhuvinod commented on clr-installer/main.go:
```diff
@@ -138,6 +139,11 @@ func execute(options args.Args) error {
 
 	if options.PamSalt != "" {
+		if status, err := user.IsValidPassword(options.PamSalt); !status {
+			fmt.Println(err)
```
Understood make sense..

Done

karthikprabhuvinod commented on user/user.go:
```diff
@@ -419,46 +416,61 @@ func writeSSHKey(rootDir string, u *User) error {
+	}
+
+	if err := validator.passwordMinimumLengthCheck(); err != nil {
+		return false, utils.Locale.Get(err.Error())
```
Done.

karthikprabhuvinod commented:
@mdhorn Addressed all the comments. Thanks for suggesting the tooltip code.

karthikprabhuvinod commented:
@mdhorn I removed the special character check as we discussed. I removed the Settooltip as we already have a password in min /max printed below text box.

mdhorn commented on locale/zh_CN/LC_MESSAGES/clr-installer.po:
```diff
@@ -812,3 +812,16 @@ msgstr "需要网络"
+msgstr "密码基于词典词"
+
+msgid "it is based on a (reversed) dictionary word"
+msgstr "La contraseña se basa en una palabra del diccionario invertida"
```
This needs to be updated for the correct locale

karthikprabhuvinod replied:
Sorry about that..done

mdhorn commented on tui/common.go:
```diff
@@ -445,20 +445,29 @@ func (page *BasePage) getModel() *model.SystemInstall {
 
 	if validation {
-		label = clui.CreateLabel(iframe, AutoSize, 1, "", Fixed)
+
```
Please remove this blank line

karthikprabhuvinod replied:
Done

mdhorn commented on gui/pages/user_add.go:
```diff
@@ -202,6 +202,7 @@ func (page *UserAddPage) onPasswordChange(entry *gtk.Entry) {
 	}
 
 	password := getTextFromEntry(page.password)
+
```
Either update the copyright notice to 2020 or better yet, revert adding this blank line and leave the file unchanged.

karthikprabhuvinod replied:
Done

karthikprabhuvinod commented:
@mdhorn : All Done..Please check. The cracklib check has been moved to a cmd package. Also added a test for it to always check if the executable exist

karthikprabhuvinod commented:
@mdhorn : Nice...I will push once more ..only updating the coverage. No, Thanks for the patience
